src/serde/de.rs

```diff
@@ -0,0 +1,997 @@
+    where
+        K: de::DeserializeSeed<'de>,
+    {
+        if self.visited {
+            return Ok(None);
+        }
+        self.visited = true;
+        seed.deserialize(DatetimeFieldDeserializer).map(Some)
+    }
+
+    fn next_value_seed<V>(&mut self, seed: V) -> Result<V::Value, IonError>
+    where
+        V: de::DeserializeSeed<'de>,
+    {
+        let timestamp = self.deserializer.reader.read_timestamp()?;
+        let datetime: DateTime<FixedOffset> = timestamp.try_into().unwrap();
+        let datetime_str = datetime.to_rfc3339();
+
+        seed.deserialize(StrDeserializer::new(Cow::from(datetime_str)))
+    }
+}
+
+struct DatetimeFieldDeserializer;
+
+impl<'de> de::Deserializer<'de> for DatetimeFieldDeserializer {
+    type Error = IonError;
+
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, IonError>
+    where
+        V: de::Visitor<'de>,
+    {
+        visitor.visit_borrowed_str("$datetime")
+    }
+
+    serde::forward_to_deserialize_any! {
+        bool u8 u16 u32 u64 i8 i16 i32 i64 f32 f64 char str string seq
+        bytes byte_buf map struct option unit newtype_struct
+        ignored_any unit_struct tuple_struct tuple enum identifier
+    }
+}
+
+struct StrDeserializer<'a> {
+    key: Cow<'a, str>,
+}
+
+impl<'a> StrDeserializer<'a> {
+    fn new(key: Cow<'a, str>) -> StrDeserializer<'a> {
+        StrDeserializer { key }
+    }
+}
+
+impl<'de> de::Deserializer<'de> for StrDeserializer<'de> {
+    type Error = IonError;
+
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, IonError>
+    where
+        V: de::Visitor<'de>,
+    {
+        match self.key {
+            Cow::Borrowed(s) => visitor.visit_borrowed_str(s),
+            Cow::Owned(s) => visitor.visit_string(s),
+        }
+    }
+
+    serde::forward_to_deserialize_any! {
+        bool u8 u16 u32 u64 i8 i16 i32 i64 f32 f64 char str string seq
+        bytes byte_buf map struct option unit newtype_struct
+        ignored_any unit_struct tuple_struct tuple enum identifier
+    }
+}
+
+pub struct StreamDeserializer<'de, R, T> {
+    de: Deserializer<R>,
+    output: PhantomData<T>,
+    lifetime: PhantomData<&'de ()>,
+}
+
+impl<'de, R, T> StreamDeserializer<'de, R, T>
+where
+    T: de::Deserialize<'de>,
+    R: StreamReader<Symbol = Symbol, Item = StreamItem>,
+{
+    pub fn new(read: R) -> Self {
+        Self {
+            de: Deserializer { reader: read },
+            output: PhantomData,
+            lifetime: PhantomData,
+        }
+    }
+}
+
+impl<'de, R, T> Iterator for StreamDeserializer<'de, R, T>
+where
+    R: StreamReader<Symbol = Symbol, Item = StreamItem>,
+    T: de::Deserialize<'de>,
+{
+    type Item = IonResult<T>;
+
+    fn next(&mut self) -> Option<IonResult<T>> {
+        // Skip any nothing
+        while let StreamItem::Nothing = self.de.reader.current() {
+            match self.de.reader.next() {
+                Ok(_) => (),
+                Err(_) => return None,
```

Comment:
I wonder, do the implementations of `StreamReader::next()` actually return an error when at the end of file? Or do they just return `Nothing`? I'm just wondering if this will ever terminate, or do we get stuck in an infinite loop of `Nothing`...

Reply:
The next should error out on end of file (hopefully)...

Reply:
Cause there is no is_end() indicator or something else so i assumed the behavior (which could be a downfall)

Reply:
`next()` returns `Nothing` if there are no more items at the current level of depth.